adapters/python/sources/message_dump_player.py

```diff
@@ -0,0 +1,162 @@
+
+    def read(self) -> Optional[Tuple[int, str, Message, bytes]]:
+        """Unpack the next message from the file."""
+
+        if not self._unpacker:
+            return None
+        message = None
+        while message is None:
+            try:
+                message = self._unpacker.unpack()
+            except msgpack.OutOfData:
+                if not self._next_message_dump_file():
+                    return None
+        ts, topic, meta, content = message
+
+        return ts, bytes(topic).decode(), load_message_from_bytes(meta), content
+
+    def _next_message_dump_file(self):
+        """Open the next message dump file from the list."""
+
+        if self._message_dump_file:
+            self._message_dump_file.close()
+
+        file_path = self._list_file.readline()
+        if not file_path:
+            return False
+
+        self._message_dump_file = open(file_path.strip(), 'rb')
+        self._unpacker = msgpack.Unpacker(self._message_dump_file)
+
+        return True
+
+    def __del__(self):
+        if self._message_dump_file:
+            self._message_dump_file.close()
+        if self._list_file:
+            self._list_file.close()
+
+
+class Egress(BaseThreadWorker):
+    """Receives messages from the dump and sends them to ZeroMQ socket."""
+
+    def __init__(
+        self,
+        reader: MessageDumpReader,
+        config: Config,
+    ):
+        super().__init__(
+            thread_name='Egress',
+            logger_name=f'{LOGGER_NAME}.{self.__class__.__name__}',
+            daemon=True,
+        )
+        self._reader = reader
+
+        config_builder = WriterConfigBuilder(config.zmq_endpoint)
+        self._writer = BlockingWriter(config_builder.build())
+        self._sync_output = config.sync_output
+        self._previous_ts = None
+
+    def workload(self):
+        self.logger.info('Starting Egress')
+        self._writer.start()
+        while self.is_running:
+            try:
+                message = self._read_message()
+                if message is not None:
+                    self.logger.debug('Sending message to the sink ZeroMQ socket')
+                    self._writer.send_message(*message)
+                else:
+                    self.logger.info('There are no more messages to send. Stopping the adapter.')
+                    self.stop()
+                    break
+            except Exception as e:
+                self.logger.error('Failed to send message: %s', e)
+                self.stop()
+                break
+        self._writer.shutdown()
+        self.logger.info('Egress was stopped')
+
+    def _read_message(self) -> Optional[Tuple[str, Message, bytes]]:
+        """Read the next message from the dump file."""
+
+        message = self._reader.read()
+
+        if message is None:
+            return None
+
+        ts, topic, meta, content = message
+
+        if self._sync_output:
+            if self._previous_ts is not None:
```

Comment:
It can work incorrectly (slower than expected) because you need to count the time you spend on reading data and sending data (actual work). Basically, if you sent m1 at 'now' and know you need to send the next one at 'now + X' then you sleep only 'now + X - new_now'. I think you understand the idea.